src/toolkits/activity_classification/activity_classifier.cpp

```diff
@@ -464,6 +464,127 @@ gl_sframe activity_classifier::predict_per_window(gl_sframe data,
   return result;
 }
 
+gl_sframe activity_classifier::predict_topk(gl_sframe data,
+                                            std::string output_type, size_t k,
+                                            std::string output_frequency) {
+  // check valid input arguments
+  if (output_type.empty()) {
+    output_type = "probability";
+  } else if (output_type != "probability" && output_type != "rank") {
+    log_and_throw(output_type +
+                  " is not a valid option for output_type.  "
+                  "Expected one of: probability, rank");
+  }
+  if (output_frequency.empty()) {
+    output_frequency = "per_row";
+  } else if (output_frequency != "per_row" &&
+             output_frequency != "per_window") {
+    log_and_throw(output_frequency +
+                  " is not a valid option for output_frequency.  "
+                  "Expected one of: per_row, per_window");
+  }
+
+  // data inference
+  std::unique_ptr<data_iterator> data_it =
+      create_iterator(data, false, false, false);
+  gl_sframe raw_preds_per_window = perform_inference(data_it.get());
+
+  // argsort probability to get index (rank) for top k class
+  flex_list class_labels = read_state<flex_list>("classes");
+  k = std::min(k, class_labels.size());
+  auto argsort_prob = [=](const flexible_type& ft) {
+    const flex_vec& prob_vec = ft.get<flex_vec>();
+    std::vector<size_t> index_vec(prob_vec.size());
+    std::iota(index_vec.begin(), index_vec.end(), 0);
+    auto compare = [&](const size_t& i, const size_t& j) {
+      return prob_vec[i] > prob_vec[j];
+    };
+    std::sort(index_vec.begin(), index_vec.end(), compare);
+    return flex_list(index_vec.begin(), index_vec.begin() + k);
+  };
+  raw_preds_per_window.add_column(
+      raw_preds_per_window["preds"].apply(argsort_prob, flex_type_enum::LIST),
+      "rank");
+
+  // get top k class name
+  size_t rank_column_index = raw_preds_per_window.column_index("rank");
+  auto get_class_name = [=](const sframe_rows::row& row) {
+    const flex_list& rank_list = row[rank_column_index];
+    flex_list topk_class;
+    for (auto i : rank_list) {
```

Comment:
Could we avoid using `auto` here? See https://google.github.io/styleguide/cppguide.html#Type_deduction

"Use type deduction only if it makes the code clearer to readers who aren't familiar with the project, or if it makes the code safer. Do not use it merely to avoid the inconvenience of writing an explicit type."

Reply:
I think auto is used here is because it is a lambda function?
https://en.cppreference.com/w/cpp/language/lambda

Reply:
This usage of `auto` is unrelated to the lambda that encloses it.

In particular, why can't this just be
```
for (const flexible_type& i : rank_list)
```

Although usually we don't use `i` for feature values. More common in our code is `ft` for a `flexible_type` value

Reply:
Okay! I thought @shantanuchhabra  is talking about the auto for the lambda function :)